src/kernel.ts

```diff
@@ -1,1069 +1,803 @@
+
+  /**
+   * Set when the message is done.  
+   */
+  isDone: boolean;
 
-    _on_error(error) {
-        /**
-         * Handle a failed AJAX request by logging the error message, and
-         * then optionally calling a provided callback.
-         *
-         * @function _on_error
-         * @param {function} error - callback
-         */
-        return (xhr, status, err) => {
-            utils.log_ajax_error(xhr, status, err);
-            if (error) {
-                error(xhr, status, err);
-            }
-        };
-    }
+}
 
-    _kernel_created(data): void {
-        /**
-         * Perform necessary tasks once the kernel has been started,
-         * including actually connecting to the kernel.
-         *
-         * @function _kernel_created
-         * @param {Object} data - information about the kernel including id
-         */
-        this.id = data.id;
-        this.kernel_url = utils.url_join_encode(this.kernel_service_url, this.id);
-        this.start_channels();
-    }
 
-    _kernel_connected(): void {
-        /**
-         * Perform necessary tasks once the connection to the kernel has
-         * been established. This includes requesting information about
-         * the kernel.
-         *
-         * @function _kernel_connected
-         */
-        events.trigger('kernel_connected.Kernel', {kernel: this});
-        // get kernel info so we know what state the kernel is in
-        var that = this;
-        this.kernel_info((reply?) => {
-            this.info_reply = reply.content;
-            events.trigger('kernel_ready.Kernel', {kernel: that});
+/**
+ * A Kernel class to communicate with the Python kernel. This
+ * should generally not be constructed directly, but be created
+ * by the `Session` object. Once created, this object should be
+ * used to communicate with the kernel.
+ */
+export
+class Kernel {
+
+  static statusChange = new Signal<Kernel, string>();
+
+  constructor(kernel_service_url: string, ws_url: string, name: string) {
+    this._id = null;
+    this._name = name;
+    this._ws = null;
+
+    this._kernelServiceUrl = kernel_service_url;
+    this._kernelUrl = null;
+    this._wsUrl = ws_url || utils.getBodyData('wsUrl');
+    if (!this._wsUrl) {
+      // trailing 's' in https will become wss for secure web sockets
+      this._wsUrl = location.protocol.replace('http', 'ws') + "//" + location.host;
+    }
+
+    this._username = "username";
+    this._sessionId = utils.uuid();
+    this._infoReply = {}; // kernel_info_reply stored here after starting
+    this._handlerMap = new Map<string, KernelFutureHandler>();
+
+    if (typeof WebSocket === 'undefined') {
+      alert('Your browser does not have WebSocket support, please try Chrome, Safari, or Firefox ≥ 11.');
+    }
+
+    this._autorestartAttempt = 0;
+    this._reconnectAttempt = 0;
+    this._reconnectLimit = 7;
+  }
+
+  /**
+   * GET /api/kernels
+   *
+   * Get the list of running kernels.
+   */
+  list(): Promise<any> {
```

Comment:
This feels really weird to have a function that is not specific to this kernel (this function talks to the server to get a list of all kernels).  @minrk, what do you think?  I think it would be better to have this function as a separate function, not in the Kernel class.